### PR TITLE
[IMP] Add to Schemas a convenience Odoo env

### PR DESCRIPTION
**Pull request by kevinkhao**
This adds a convenience Env to the Schemas behind the datamodels.

Benefits include:
- easily being able to keep all validation logic on datamodels, in consistent style with Odoo. For example, it is very easy to check values in @validates functions against in-db values 
- you could create a new record from validated data in a @post_load function

Potential drawbacks:
- the env is made with superuser id. Wrong or convoluted usage can lead to suspension of access rules
- adds modification on the Schema level, might be a cleaner way (?)

### CODE REVIEW

**Comment by lmignon** [comment]
@kevinkhao Thank you for the contrib! IMO we should find a way to have an env initialized with the current user.... (I've to think about it but it should be possible since it's what odoo is doing)

**Comment by kevinkhao** [comment]
@lmignon would be very interested if you have an idea!

File: requirements.txt [hunk 1]
```diff
@@ -4,6 +4,6 @@ parse-accept-language
 jsondiff
 graphene
 graphql-server-core
-marshmallow
+marshmallow==3.2.2
```
Comment (lmignon) [inline requirements.txt:7]: PLZ don't freeze to a specific version. If 3.2.2 is the minimal version required you can set `marshmallow>=3.2.2` If you have some troubles with newer version can you describe the problem and the marshmallow version?

**Comment by lmignon** [comment]
@kevinkhao I've to dig into odoo... No idea for the moment and a lot of work..

**Comment by kevinkhao** [comment]
@lmignon Here is some [code](https://github.com/akretion/sale-import/blob/master/sale_import_base/datamodels/sale_order.py) I worked on (it's old and very WIP but principles are visible).
What do you think, is this what use you envisioned first ?
Two things to note:
- line 16 function _validate_address_customer, having an .env allows validation against database data and expands validation possibilities instead of having some simple validation
- line 32 format_vals , I use this to do some transformations on the data so that I can just do a 
`vals = my_datamodel.load(my_data_to_be_validated).dump()
new_sale_order = self.env['sale.order'].create(vals)`
the idea is to have load() try to do a mapping that's immediately usable in odoo (in principle, no direct object modification)

So, same question about it, what do you think of this organization, did you envision something similar ?

**Comment by lmignon** [comment]
@kevinkhao The possibilities offered by being able to get access to `env` on the datamodel itself are great. We can threrefore consider that the property `env` on the datamodel will be part of the API. My only concern is to find a way to have an env initialized properly (with the current user into the current transaction). I really hope to find some times to dig into odoo to improve your proposal.

**Comment by lmignon** [comment]
@kevinkhao the link to your sample code is broken....

**Comment by lmignon** [comment]
@kevinkhao I just fixed a compatibility issue with the latest release of marshmallow (https://github.com/OCA/rest-framework/pull/48/commits/fd782728f3287aa8c0d0061ee25bc703cc2e75b0) and found a way to get access to the current Odoo Environment from the datamodel instance and the schema  https://github.com/OCA/rest-framework/pull/48/commits/7027645a905f06bcc96f5f7d3cf3d49ce03dde9b
Feedback welcome!

File: datamodel/core.py [hunk 2]
```diff
@@ -343,6 +348,15 @@ def _build_datamodel(cls, registry):
 
         return DatamodelClass
 
+    @classmethod
+    def _get_env_from_registry(cls, registry):
```
Comment (lmignon) [inline datamodel/core.py:352]: @kevinkhao Your proposal doesn't work. Indeed the env from datamodels will be into an other transaction and with the wrong user.....

File: requirements.txt (hunk 1 above)
Comment (lmignon) [reply requirements.txt:7]: @kevinkhao The problem is now solved. Freeze is no more required.

**Comment by lmignon** [comment]
@kevinkhao I close this one since the functionality has been implemented in another way into OCA@7027645 Feel free to comment these last changes

**Pull request closed without merging**